examples/toy/Ch4/mlir/ToyCombine.cpp

```diff
@@ -21,135 +21,78 @@
 //===----------------------------------------------------------------------===//
 
 #include "toy/Dialect.h"
-
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Pass/Pass.h"
 #include "mlir/IR/Matchers.h"
 #include "mlir/IR/PatternMatch.h"
 #include "mlir/IR/StandardTypes.h"
-
 #include <numeric>
+using namespace mlir;
+using namespace toy;
 
-namespace toy {
-
-namespace {
-
-/// Fold transpose(transpose(x) -> transpose(x)
-struct SimplifyRedundantTranspose : public mlir::OpRewritePattern<TransposeOp> {
-  /// We register this pattern to match every toy.transpose in the IR.
-  /// The "benefit" is used by the framework to order the patterns and process
-  /// them in order of profitability.
-  SimplifyRedundantTranspose(mlir::MLIRContext *context)
-      : OpRewritePattern<TransposeOp>(context, /*benefit=*/1) {}
-
-  /// This method is attempting to match a pattern and rewrite it. The rewriter
-  /// argument is the orchestrator of the sequence of rewrites. It is expected
-  /// to interact with it to perform any changes to the IR from here.
-  mlir::PatternMatchResult
-  matchAndRewrite(TransposeOp op,
-                  mlir::PatternRewriter &rewriter) const override {
-    // Look through the input of the current transpose.
-    mlir::Value *transposeInput = op.getOperand();
-    TransposeOp transposeInputOp =
-        llvm::dyn_cast_or_null<TransposeOp>(transposeInput->getDefiningOp());
-    // If the input is defined by another Transpose, bingo!
-    if (!transposeInputOp)
-      return matchFailure();
-
-    // Use the rewriter to perform the replacement
-    rewriter.replaceOp(op, {transposeInputOp.getOperand()}, {transposeInputOp});
-    return matchSuccess();
-  }
-};
-
-/// Fold reshape(constant(x)) -> constant(x'), with x' being reshaped in place.
-struct SimplifyReshapeConstant : public mlir::OpRewritePattern<ReshapeOp> {
-  using mlir::OpRewritePattern<ReshapeOp>::OpRewritePattern;
-
-  mlir::PatternMatchResult
-  matchAndRewrite(ReshapeOp reshape,
-                  mlir::PatternRewriter &rewriter) const override {
-    // Look through the input of the current reshape.
+// Helper function to fold reshape(constant) in place
+Value *reshapeConstant(Builder &builder, Value* arg) {
+    ReshapeOp reshape = llvm::dyn_cast_or_null<ReshapeOp>(arg->getDefiningOp());
+    mlir::OpBuilder builder2(reshape.getOperation());
     ConstantOp constantOp = llvm::dyn_cast_or_null<ConstantOp>(
         reshape.getOperand()->getDefiningOp());
-    // If the input is defined by another constant, bingo!
-    if (!constantOp)
-      return matchFailure();
-
-    auto reshapeType = reshape.getType().cast<ToyArrayType>();
+    auto reshapeType = reshape.getType().cast<TensorType>();
     if (auto valueAttr =
```

Comment:
This is the only branch that is possible. All of the toy operations work exclusively on DenseElementsAttr. All this function should need to do is something like:

DenseElementsAttr reshapeConstant(DenseElementsAttr attr, Value *arg) {
  return attr.reshape(arg->getType());
}